diag_manager/fms_doubly_linked_list.F90

```diff
@@ -0,0 +1,275 @@
+!> \author Miguel Zuniga
+!> \email miguel.zuniga@noaa.gov
+!! \brief A module for a double linked-list.
+!!
+!! \description A double linked list module that is rougly a fortran translation of the C++
+!! boubly linked list from the book  ``Data Structures And Algorithm Analysis in C++", 3rd Edition,
+!!  by Mark Allen Weiss
+module fms_doubly_linked_list_mod
+  implicit none
+
+  !> Linked list node
+  type, private :: node
+    private
+    class(*), pointer :: data => null()
+    type(node), pointer :: next => null()
+    type(node), pointer :: prev => null()
+  end type node
+
+  !> Linked list literator
+  type, public :: literator
+    type(node), pointer :: current
+    type(node), pointer :: end
+  contains
+    procedure :: has_data => literator_has_data
+    procedure :: next => literator_next
+    procedure :: get => literator_data
+  end type literator
+
+  type, public :: linked_list
+    !! Note we are overriding the default constructor with an
+    ! interface of the same name
+    type(node), pointer :: head
+    type(node), pointer :: tail
+    integer :: theSize
+  contains
+    procedure :: push_front => push_at_front
+    procedure :: push_back => push_at_back
+    procedure :: pop_front => pop_at_front
+    procedure :: pop_back => pop_at_back
+    procedure :: insert => insert_data
+    procedure :: remove => remove_node
+    procedure :: get_literator => get_forward_literator
+    procedure :: size => get_size
+    procedure :: is_empty => is_size_zero
+    procedure :: clear => clear_all
+    final :: destructor
+  end type linked_list
+
+  interface node
+    module procedure :: node_constructor
+  end interface node
+
+  interface linked_list
+    module procedure :: linked_list_constructor
+  end interface linked_list
+
+  interface literator
+    module procedure :: literator_constructor
+  end interface literator
+
+contains
+
+  !> Insert data d in a new node to be placed in front of the
+  ! target node t_nd. Return an iterator that starts with the
+  ! newly inserted node.
+  function  insert_data( this, t_nd,  d ) result(liter)
+    class(linked_list), intent(in out) :: this
+    class(node), pointer, intent (in)  :: t_nd
+    class(*), target, intent(in)       :: d
+    class(literator), allocatable      :: liter
+    !!
+    class(node), pointer :: nd
```

Comment:
Variables need to be documented in doxygen style

Reply:
Done.